### PR TITLE
Create general PoolingOp and decompose to MaxPool2d

### PR DESCRIPTION
**Pull request by LPanosTT**
- Added a new PoolingOp that can be represent any kind of pooling (max, min, avg, etc) with any dimensionality.
 - Added a decomposition pass that can fish out whether or not a PoolingOp corresponds to a MaxPool2d and rewrites it.
 - @nsmithtt Removed pass that adds reshapes on the input and output of `ttir.max_pool2d`, the reshapes are now inserted upon conversion to `ttnn.max_pool2d`

The PoolingOp is meant to closely resemble the `stablehlo.reduce_window` op. One of its features is that it can take any number of inputs and Pool them all simultaneously, and return a tuple with all the outputs. In the decomposition pass I convert this into multiple individual maxpool2d ops if need be. There is a test for this.

### CODE REVIEW

File: include/ttmlir/Dialect/TTIR/IR/TTIROpsEnums.td [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+#define TTMLIR_TTIR_ENUMS_TD
+
+include "mlir/IR/EnumAttr.td"
+
+def TTIR_AveragePoolingMethod : I32EnumAttrCase<"Average", 0>;
+def TTIR_MaxPoolingMethod : I32EnumAttrCase<"Max", 1>;
+
+def TTIR_PoolingMethod : I32EnumAttr<"PoolingMethod", "TTIR PoolingMethod", [
+                                      TTIR_AveragePoolingMethod,
+                                      TTIR_MaxPoolingMethod
```
[the remark this answers is not in the document]
Comment (LPanosTT) [reply include/ttmlir/Dialect/TTIR/IR/TTIROpsEnums.td:15]: Sure

File: include/ttmlir/Dialect/TTIR/IR/TTIROps.td [hunk 2]
```diff
@@ -728,9 +755,7 @@ def TTIR_MaxPool2dOp : TTIR_DPSOp<"max_pool2d"> {
                          SI32Attr:$padding_right,
                          SI32Attr:$padding_top,
                          SI32Attr:$padding_bottom,
-                         TT_OperandConstraintArrayAttr:$operand_constraints,
-                         OptionalAttr<SI32Attr>:$original_height,
-                         OptionalAttr<SI32Attr>:$original_width);
```
Comment (nsmithtt) [inline include/ttmlir/Dialect/TTIR/IR/TTIROps.td:733]: Thank you!

File: lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp [hunk 3]
```diff
@@ -501,105 +565,60 @@ class StableHLOToTTIRReduceWindowOpConversionPattern
...
+
+    auto operand_constraints = rewriter.getArrayAttr(SmallVector<Attribute>(
+        adaptor.getOperands().size(), rewriter.getAttr<OperandConstraintAttr>(
+                                          OperandConstraint::AnyDeviceTile)));
+
+    mlir::tt::ttir::PoolingMethod pooling_method;
+    if (isMaxPool(srcOp)) {
+      pooling_method = mlir::tt::ttir::PoolingMethod::Max;
+    } else {
+      return failure();
```
Comment (nsmithtt) [inline lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp:613]: If we make it this far down in the rewriter, maybe we should fail with an error message: `rewriter.notifyMatchFailure(op, "Unsupported pooling method")`.

File: lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp [hunk 4]
```diff
@@ -402,9 +403,228 @@ struct ConvolutionToConv2dPattern
...
+
+    // The indices of the true window dimensions and strides must be the same
+    if ((true_window_dimensions_indices[0] != true_window_stride_indices[0] ||
+         true_window_dimensions_indices[1] != true_window_stride_indices[1]) &&
+        (true_window_dimensions_indices[0] != true_window_stride_indices[1] ||
+         true_window_dimensions_indices[1] != true_window_stride_indices[0])) {
+      return failure();
+    }
+
+    // NOTE: To model later
```
Comment (nsmithtt) [inline lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp:469]: We should have an issue tracking todos

**Review by nsmithtt — APPROVED** [review]
Minor comments inline

File: lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp (hunk 4 above)
Comment (LPanosTT) [reply lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp:469]: I think Im going to go more with the rewrite failure you mentioned above. I think we want to go with supporting maxpool methods as we go. As of right now I believe ttnn only supports maxpool with square windows, which is a much easier thing to look for.

**Review by jnie-TT — APPROVED** [review]
Runtime changes look good

File: lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp [hunk 5]
```diff
@@ -406,9 +407,227 @@ struct ConvolutionToConv2dPattern
...
+    // The indices of the true window dimensions and strides must be the same
+    if ((true_window_dimensions_indices[0] != true_window_stride_indices[0] ||
+         true_window_dimensions_indices[1] != true_window_stride_indices[1]) &&
+        (true_window_dimensions_indices[0] != true_window_stride_indices[1] ||
+         true_window_dimensions_indices[1] != true_window_stride_indices[0])) {
+      return failure();
+    }
+
+    // Padding must be 8 in length
+    if (op.getPadding().size() != 8) {
```
Comment (svuckovicTT) [inline lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp:474]: 2 padding values per dim = 8 values? Should their indices also match up to window dims&strides?

File: lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp [hunk 6]
```diff
@@ -406,9 +407,227 @@ struct ConvolutionToConv2dPattern
...
+    const int64_t SPATIAL_H = -3;
+    const int64_t SPATIAL_W = -2;
+    const int64_t NON_SPATIAL = -1;
+
+    auto inputType =
+        mlir::cast<RankedTensorType>(adaptor.getInputs()[0].getType());
+    assert(inputType.getRank() == 4 && "Input must be 4D tensor");
+    std::vector<int64_t> desired_layout(inputType.getRank(), NON_SPATIAL);
+    desired_layout[inputType.getRank() - 3] = SPATIAL_H;
+    desired_layout[inputType.getRank() - 2] = SPATIAL_W;
```
Comment (svuckovicTT) [inline lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp:494]: If rank is 4, that'd put H/W dims into second and third place, like this [x, H, W, x]? Oh, it's NHWC?

File: lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp [hunk 7]
```diff
@@ -406,9 +407,227 @@ struct ConvolutionToConv2dPattern
...
+
+    int64_t non_spatial_count = 0;
+    for (int64_t i = 0; i < static_cast<int64_t>(desired_layout.size()); i++) {
+      if (desired_layout[i] == NON_SPATIAL) {
+        desired_layout[i] = non_spatial_count;
+        non_spatial_count++;
+      }
+    }
```
Comment (svuckovicTT) [inline lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp:502]: Do we need to calculate this? If rank is 4, and we manually set 2 dims to spatial, this is always going to be 2?

Comment (svuckovicTT) [reply lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp:502]: We calculate this again a few lines below, but I don't think it's used anywhere?

File: lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp [hunk 8]
```diff
@@ -406,9 +407,227 @@ struct ConvolutionToConv2dPattern
...
+          padding_right_attr, operand_constraints);
+
+      Value output = generateTransposeSequence(new_pool.getResult(), rewriter,
+                                               output_transpose_indices,
+                                               operand_constraints);
+      outputs.push_back(output);
+    }
+
+    for (int64_t i = 0; i < static_cast<int64_t>(outputs.size()); i++) {
+      rewriter.replaceAllUsesWith(op.getResult(i), outputs[i]);
```
Comment (svuckovicTT) [inline lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp:581]: `replaceAllUsesWith` should not be used in dialect conversion. It's a bug that it's available to dialect conversion's rewriter. It modifies IR, which should not be allowed in dialect conversion rewriter.

File: lib/Conversion/TTIRToTTNN/CMakeLists.txt [hunk 9]
```diff
@@ -1,3 +1,4 @@
+set(CMAKE_BUILD_TYPE Debug)
```
Comment (svuckovicTT) [inline lib/Conversion/TTIRToTTNN/CMakeLists.txt:1]: remove

File: lib/Conversion/TTIRToTTNN/TTIRToTTNN.cpp [hunk 10]
```diff
@@ -503,9 +506,19 @@ class ConstantOpConversionPattern
 
     if (valueAttr.isSplat()) {
       Value device = getOrInsertDevice(rewriter, op);
-      float fillValue = valueAttr.getElementType().isInteger()
-                            ? static_cast<float>(valueAttr.getSplatValue<int>())
-                            : valueAttr.getSplatValue<float>();
+      float fillValue = 0;
+      if (op.getResult().getType().getElementType().isBF16()) {
+        uint32_t raw_shifted_bfloat = valueAttr.getSplatValue<uint32_t>() << 16;
+        fillValue = *reinterpret_cast<float *>(&raw_shifted_bfloat);
```
Comment (svuckovicTT) [inline lib/Conversion/TTIRToTTNN/TTIRToTTNN.cpp:512]: I don't understand how this works, can you add a comment to explain it please?

File: lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp [hunk 11]
```diff
@@ -146,46 +147,46 @@ generateTranspose(Value input, int64_t dim0, int64_t dim1,
       dim1_attr, operandConstraints);
 }
 
-static std::vector<TransposeDims> generateKernelTransposeIndices(
-    ttir::ConvolutionOp op,
-    const std::vector<int64_t> ttnn_convolution_kernel_layout) {
+static std::vector<TransposeDims>
+generateTransposeIndicesLeftToRight(std::vector<int64_t> current_layout,
```
Comment (svuckovicTT) [inline lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp:151]: All these transpose methods are hard to follow. A comment (with an example) above each method would make it much easier to come back to.

Additionally, it's somewhat noisy to keep these method in this file, can they be moved to a util file of sorts?

**Review by svuckovicTT — CHANGES_REQUESTED** [review]
Thanks for packing the reshapes to the conversion itself, that makes it a lot cleaner!

Requesting changes for the following reasons:
1. replaceAllUsesWith needs to be removed
2. in more than one file, I've seen a mix of camelCase and snake_case, can you please go over all the files and correct this?
3. personally, I'd like to see more comments - pooling ops are among the most complicated to work with but with properly commented code, it gets easier :)

File: lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp [hunk 12]
```diff
@@ -501,105 +565,60 @@ class StableHLOToTTIRReduceWindowOpConversionPattern
...
+                 : rewriter.getDenseI64ArrayAttr(
+                       SmallVector<int64_t>(window_dimensions.size() * 2, 1));
+
+    auto operand_constraints = rewriter.getArrayAttr(SmallVector<Attribute>(
+        adaptor.getOperands().size(), rewriter.getAttr<OperandConstraintAttr>(
+                                          OperandConstraint::AnyDeviceTile)));
+
+    mlir::tt::ttir::PoolingMethod pooling_method;
+    if (isMaxPool(srcOp)) {
+      pooling_method = mlir::tt::ttir::PoolingMethod::Max;
```
[the remark this answers is not in the document]
Comment (LPanosTT) [reply lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp:611]: No but in the future we'll have sum-pooling and avg-pooling and min-pooling, etc.

File: lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp [hunk 13]
```diff
@@ -501,105 +565,60 @@ class StableHLOToTTIRReduceWindowOpConversionPattern
...
+    auto window_strides = adaptor.getWindowStridesAttr();
+    auto base_dilations = adaptor.getBaseDilationsAttr();
+    auto window_dilations = adaptor.getWindowDilationsAttr();
+    auto padding_ = adaptor.getPaddingAttr();
+
+    // Generate defaults if they dont exist
+    window_strides =
+        window_strides ? window_strides
+                       : rewriter.getDenseI64ArrayAttr(
+                             SmallVector<int64_t>(window_dimensions.size(), 1));
```
[the remark this answers is not in the document]
Comment (LPanosTT) [reply lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp:590]: It gets set to 1 for all dimensions by default: ` SmallVector<int64_t>(window_dimensions.size(), 1)`

File: lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp (hunk 5 above)
Comment (LPanosTT) [reply lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp:474]: No because of the high and low. We might want to pad the top differently than the bottom even though they exist on the same dimension.

File: lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp (hunk 7 above)
Comment (LPanosTT) [reply lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp:502]: I'll remove this, its leftover from when I was trying out a different method of doing this.

File: lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp (hunk 8 above)
Comment (LPanosTT) [reply lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp:581]: Not sure what to do here then. The `PoolingOp` can perform multiple pools at once. That is, you could pass it some number (lets say 5) tensors, pool all of them, and then get 5 results. As part of this conversion I am splitting that up into individual MaxPool2d ops. So I need to replace the use of the individual results with each new MaxPool2d op.

Also this isn't dialect conversion its just a decomposition pass within the same dialect.

File: lib/Conversion/TTIRToTTNN/TTIRToTTNN.cpp (hunk 10 above)
Comment (LPanosTT) [reply lib/Conversion/TTIRToTTNN/TTIRToTTNN.cpp:512]: Yeah so since bfloat16 isn't a type I can just have I need to just move around the bits representing the scalar value here I get the splat value as a uint32 so that I can perform a bit shift. The bfloat16 bits will be in the right-hand 16 bits of the uint32_t. A bfloat16 is the same as a float32 but with a much smaller mantissa. So, bit shifting this 16 bits to the right performs the cast to float32. I need to use a reinterpret cast on a pointer so that the bits remain the same.

File: lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp (hunk 8 above)
Comment (LPanosTT) [reply lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp:581]: Looks like this PR will solve this issue for me actually so don't worry about my floating-point bit-shifting hack: https://github.com/tenstorrent/tt-mlir/pull/1164#pullrequestreview-2424129767

File: lib/Conversion/TTIRToTTNN/TTIRToTTNN.cpp (hunk 10 above)
Comment (LPanosTT) [reply lib/Conversion/TTIRToTTNN/TTIRToTTNN.cpp:512]: Looks like this PR will solve this issue for me actually so don't worry about my floating-point bit-shifting hack: https://github.com/tenstorrent/tt-mlir/pull/1164

File: lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp (hunk 7 above)
Comment (LPanosTT) [reply lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp:503]: Actually my mistake. I do need these. I want to assign the non-spatial dims different values in the same order from left to right so that we can accurately generate the correct transposes to make the current layout match the desired layout.

File: lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp (hunk 11 above)
Comment (LPanosTT) [reply lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp]: I've cleaned them up and removed a number of them. Also added comments.

File: lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp (hunk 8 above)
Comment (LPanosTT) [reply lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp]: I've resolved this. I didn't realize I could use `replaceOp` with an op and range of results.

File: lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp [hunk 14]
```diff
@@ -497,11 +498,72 @@ class StableHLOToTTIRReduceWindowOpConversionPattern
...
+          return false;
+        }
+      } else if (init_value_op.getValue().getType().isInteger(64)) {
+        if (*init_value_op.getValue().value_begin<int64_t>() !=
+            std::numeric_limits<int64_t>::min()) {
+          return false;
+        }
+      } else {
+        return false;
+      }
```
Comment (svuckovicTT) [inline lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp:564]: Consider packing into its own method.

File: lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp [hunk 15]
```diff
@@ -115,77 +117,77 @@ enum ConvolutionKernelDimension {
...
-}
+/*
+ * This function will use a sequence of transpose indices to
+ * generate the actual transpose operations descrbibed by them.
+ *
+ * It takes an input to apply these transposes to and returns the
+ * result at the end of the sequence
+ */
+static Value
+generateTransposeSequence(Value input, PatternRewriter &rewriter,
```
Comment (svuckovicTT) [inline lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp:156]: Consider renaming `generateTransposeSequence` to `applyTransposeIndices`.

File: lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp [hunk 16]
```diff
@@ -115,77 +117,77 @@ enum ConvolutionKernelDimension {
...
+                          std::vector<TransposeDims> transpose_indices,
+                          ::mlir::ArrayAttr operand_constraints) {
+  for (auto [dim0, dim1] : transpose_indices) {
 
-static std::vector<TransposeDims> generateKernelTransposeIndices(
-    ttir::ConvolutionOp op,
-    const std::vector<int64_t> ttnn_convolution_kernel_layout) {
-  std::vector<TransposeDims> transpose_indices;
+    auto input_type = mlir::cast<mlir::RankedTensorType>(input.getType());
```
Comment (svuckovicTT) [inline lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp:161]: Still lots of snake_case, can you please go over all the changes and adjust?

File: lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp (hunk 5 above)
Comment (svuckovicTT) [reply lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp:438]: Right, 2-sided padding for each dim, just wanted to clarify that. The second question was whether we should have the same test that we do above, but for padding as well, so that we confirm they "see" the same dimensions as spatial.

File: lib/Conversion/TTIRToTTNN/TTIRToTTNN.cpp [hunk 17]
```diff
@@ -23,6 +23,9 @@
 #include "llvm/Support/Casting.h"
 #include "llvm/Support/ErrorHandling.h"
 #include "llvm/Support/LogicalResult.h"
+#include <mlir/IR/Attributes.h>
+#include <mlir/IR/BuiltinAttributes.h>
+#include <mlir/IR/Value.h>
```
Comment (svuckovicTT) [inline lib/Conversion/TTIRToTTNN/TTIRToTTNN.cpp:28]: I don't know why VSCode (or whichever plugin) includes these with `<>` by default, but it should be `""` as these are not standard libs like cstdio, algorithm, ...

Copy pasting a snippet on include styles - maybe not applicable for this specific file, but I saw some files above that would benefit from following this:

Include style is defined by LLVM coding standards: https://llvm.org/docs/CodingStandards.html#include-style

We prefer these #includes to be listed in this order:

	1. Main Module Header
	2. Local/Private Headers
	3. LLVM project/subproject headers (clang/..., lldb/..., llvm/..., etc)
	4. System #includes

Using `TTIRToTTNN.cpp` as an example, this is what includes would look like for us:

```
#include "ttmlir/Conversion/TTIRToTTNN/TTIRToTTNN.h"  # main header

#include "ttmlir/Dialect/TT/IR/TTOpsTypes.h"  # these are local/private headers
#include "ttmlir/Dialect/TTNN/Utils/Utils.h"

#include "mlir/Dialect/MemRef/IR/MemRef.h"  # llvm project/subproj headers
#include "llvm/Support/LogicalResult.h"

#include <cstdio>  # system includes
#include <algorithm>
```

File: lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp (hunk 8 above)
Comment (svuckovicTT) [reply lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp]: > Also this isn't dialect conversion its just a decomposition pass within the same dialect.

True, but we are actually using dialect conversion for the decomp, as it offers op legality checks on target IR.

**Review by svuckovicTT — APPROVED** [review]
Thanks for adding comments Lewis, I think it's much easier to follow what's going on now!

Approving, but please make sure to change snake_case to camelCase before merging.

File: lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp (hunk 15 above)
Comment (LPanosTT) [reply lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp:156]: I think `generateTransposeOps` might be better

File: lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp (hunk 14 above)
Comment (LPanosTT) [reply lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp:564]: Will do

File: lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp (hunk 16 above)
Comment (LPanosTT) [reply lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp:161]: Yes
